chore(docs): wget reference docs (#3205)
open-component-model/open-component-model: website/content/docs/how-to/add-resources-from-http-urls.md

Reviewer: I believe most of our How-Tos give the user a reproducible task they can follow. This How-To feels more like a general help-page what is possible with the `wget` resource type. Would it make sense to think about a real scenario and lead the user through? WDYT?

Author: Maybe I should rewrite this more like "migration from ocmv1 wget"

Author: but I do not want to :D

Reviewer: This How-To contains the several sections:
- The actual steps to "Add resources from HTTP URLs"
- Setting the media type
- Authenticating against a protected server
- Pinning an expected digest
- Sending a non-Get request
- Tuning the download
- Migrating from OCM v1

If you have a look at the other How-Tos on our website, then they usually have
- Steps
- Rarely "Tips"
- Rarely another section explaining the concept/background

I think a lot of the stuff you have written down here is important but it is a bit to much for a How-To and it is loosely coupled. For a user who just wants to use that resource and see a small example on how to do so, it feels too much.

Maybe a tutorial would be a better idea, where we can show why it makes sense to pin the expected digest and configure the download etc.pp while following the steps.

Reviewer 2: I second @frewilhelm comment. This is way too much for a simple hopw-to. We should reduce the content to just the current steps plus the media type thingy as this required for the specified resource. I would even cur out the input type and just mention input types and reference the input-and-access-types reference doc. All the other details and the input type usage should go into a tutorial which could look pretty much the same as this current how-to.

        type: Wget/v1
        url: https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar
      digest:
        hashAlgorithm: SHA-256
        normalisationAlgorithm: genericBlobDigest/v1
        value: 6e3205bbad194f902ee8bdc5a712c47f7a5443fabfd066ef1e95088c837fe0ae
      name: ocm-cli
      relation: external
      type: blob
      version: 1.0.0
```

{{< /details >}}

In both cases, the resource has the same digest, computed over the bytes that were fetched.

{{< /step >}}

{{< step >}}

### Download the resource back

This proves the transport (and, on a protected server, the credentials) work end-to-end:

```bash
ocm download resource ./transport-archive//github.com/acme.org/myapp:1.0.0 \
  --identity name=ocm-cli \
  --output ./ocm-cli
```

You should see: `level=INFO msg="resource downloaded successfully" output=./ocm-cli`.

{{< callout context="note" >}}
When the media type is an archive the CLI can unpack, `--output` is a **directory** with the extracted contents.
Other media types are written to a file at that path. `cli.tar` is a `.tar`, so `./ocm-cli` is a directory here.
{{< /callout >}}

{{< /step >}}

{{< /steps >}}

> ## Authenticate against a protected server

Reviewer: I don't understand why this section is in the How-To. I would've expected it in the tutorial

Reviewer 3: It's literally a How to authenticate against a protected server... why would this be in the tutorial that expected you to read the "how to do things"?

---
title: "Add Resources from HTTP URLs"
description: "Add a file served over HTTP or HTTPS to a component version with the Wget input or access type."
icon: "🌐"
weight: 17
toc: true
---

## Goal

Add a file served over plain HTTP or HTTPS (a release archive, a checksum, a signed binary) to a component version
using the `Wget/v1` type.

> OCM has two options where the difference is only the end location:

Reviewer 2: ```suggestion
OCM provides two types for HTTP resources. They differ in when the file is fetched and where the bytes end up:
```

- The [**input type**]({{< relref "docs/reference/input-and-access-types.md#wgetv1-input" >}}) downloads the file while
  the component version is built and stores it as a localBlob in the component version.
- The [**access type**]({{< relref "docs/reference/input-and-access-types.md#wgetv1-access" >}}) stores only the
  **URL** and leaves the file on the remote server.

> Both use the same download and credential code, so everything below applies to either one. For the reasoning behind

Reviewer 2: ```suggestion
Both use the same download and credential code, so everything below applies to either one. For the full set of options and the reasoning behind them, see the tutorial
```

> each option and the full set of variable, see the tutorial

Reviewer 2: ```suggestion
Both types share the same download and credential code, so the rest of this guide applies to either one. For the full set of options and the reasoning behind them, see the tutorial 
```

Reviewer 4: > Both types share the same download and credential code

maybe rather "Both types share the same download and credential functionality"

[Work with HTTP Resources]({{< relref "docs/tutorials/wget-http-resources.md" >}}).

## You'll end up with

> - A component version containing a resource fetched over HTTP or HTTPS, either embedded as a local blob or referenced by URL

Reviewer 2: ```suggestion
- A component version containing a resource fetched over HTTP or HTTPS, either embedded as a local blob or referenced by URL.
```

> - That same resource downloaded back out, to confirm the round trip works

Reviewer 2: ```suggestion
- The same resource retrieved from the archive, verifying the round trip end to end.
```

Reviewer 3: >retrieved from the archive

What archive? 🤔

**Estimated time:** ~10 minutes

## Prerequisites

- [OCM CLI]({{< relref "docs/getting-started/ocm-cli-installation.md" >}}) installed
- A working directory for the constructor and the transport archive

> This guide fetches a public OCM release asset, so it runs end to end with no registry and no credentials. To point it

Reviewer 2: ```suggestion
This guide fetches a public OCM release asset, so it runs end-to-end with no registry access or credentials required. To point it
```

Reviewer 3: >registry access or credentials

These are the same thing. 🤔

at a server that needs authentication, see [Authenticate against a protected server](#authenticate-against-a-protected-server).

## Steps

{{< steps >}}

{{< step >}}

> ### Choose the input type or the access type

Reviewer 2: ```suggestion
### Choose input or access type
```

> ### Choose the input type or the access type

Reviewer 2: ```suggestion
### Choose the input type or the access type

Pick either one to follow this guide:
```

> | Choose the **input type** when                     | Choose the **access type** when                    |

Reviewer 2: ```suggestion
| Choose **input type** when                     | Choose **access type** when                    |
```

Reviewer 2: just to not repeat the "choose" from the step title

|----------------------------------------------------|----------------------------------------------------|
| The file must stay reproducible and work offline   | The URL is the source of truth and should stay so  |
| The URL might disappear or change what it serves   | The file is large and you don't want to copy it    |

> | You are building an air-gapped repository          | Consumers should fetch directly from the origin    |

Reviewer 4: I don't think this is the right place for this comparison. I think we should either add something like this to the `input and access types` reference docs or even add a completely separate tutorial (when to use input or acess type?)

Reviewer 3: I can remove this entire thing then. :)

> Pick either one to follow this guide.

Reviewer 2: ```suggestion
```

Reviewer 2: move this up ^^ right behind the step title

{{< callout context="caution" >}}

> A Wget resource is transferred **by value**. When you move the component version with `ocm transfer cv`, pass

Reviewer 2: ```suggestion
A Wget resource is transferred **by value**. When you move the component version with `ocm transfer cv` and pass the 
```

> A Wget resource is transferred **by value**. When you move the component version with `ocm transfer cv`, pass

Reviewer 4: > A Wget resource is transferred **by value**.

Not sure what message you are trying to convey with this "caution"? I mean, it is only transferred by value IF you set --copy-resources.

Reviewer 3: Ah, I think I messed this up here. The by value was that it's always downloaded but this is regarding transfer. So yeah.

> `--copy-resources true`; the `Wget/v1` access is then replaced by a

Reviewer 2: ```suggestion
`--copy-resources` flag; the `Wget/v1` access is then replaced by a
```

[`LocalBlob/v1`]({{< relref "docs/reference/input-and-access-types.md#localblobv1" >}}) in the target. See
[Transfer behavior]({{< relref "docs/tutorials/wget-http-resources.md#how-transfer-works" >}}) for more details.
{{< /callout >}}

{{< /step >}}

{{< step >}}

### Create the component constructor

> Both spellings describe the same resource with the same fields. Write one of them to `component-constructor.yaml`:

Reviewer 2: ```suggestion
Both spellings below describe the same resource with the same fields. Write one of them to `component-constructor.yaml`:
```

> Both spellings describe the same resource with the same fields. Write one of them to `component-constructor.yaml`:

Reviewer 4: ```suggestion
Both variants describe the same resource with the same fields. Write one of them to `component-constructor.yaml`:
```
spellings?

```bash
cat > component-constructor.yaml << 'EOF'
components:
  - name: github.com/acme.org/myapp
    version: 1.0.0
    provider:
      name: acme.org
    resources:
      - name: ocm-cli
        type: blob
        version: 1.0.0
        input:
          type: Wget/v1
          url: https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar
          mediaType: application/x-tar
EOF
```

{{< /tab >}}
{{< tab "Access type" >}}

```bash
cat > component-constructor.yaml << 'EOF'
components:
  - name: github.com/acme.org/myapp
    version: 1.0.0
    provider:
      name: acme.org
    resources:
      - name: ocm-cli
        type: blob
        version: 1.0.0
        relation: external
        access:
          type: Wget/v1
          url: https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar
          mediaType: application/x-tar
EOF
```

{{< /tab >}}
{{< /tabs >}}

Set `mediaType` explicitly here because GitHub serves every release asset as `application/octet-stream`. If you leave
it out, OCM falls back to the server's `Content-Type`, then to `application/octet-stream`. The full order is in

> [Set the media type]({{< relref "docs/tutorials/wget-http-resources.md#set-the-media-type" >}}).

Reviewer 4: ```suggestion
Set `mediaType` explicitly here. Otherwise, OCM will default to the `Content-Type` served by GitHub, which for release assets is always `application/octet-stream`. The full `mediaType` defaulting mechanism of `Wget` is described in
[Set the media type]({{< relref "docs/tutorials/wget-http-resources.md#set-the-media-type" >}}).
```

{{< callout context="caution" >}}

> **Never put credentials in `url`, `header`, or `body`.** That includes userinfo (`https://user:token@host/...`) and

Reviewer 2: ```suggestion
**Never put credentials in `url`, `header`, or `body`.** That includes user info (`https://user:token@host/...`) and
```

> presigned query parameters. An access specification is stored in the component descriptor, signed, and is included in

Reviewer 2: ```suggestion
presigned query parameters. An access specification is stored in the component descriptor then signed and finally included in
```

the transfer. An input specification lives in your constructor file, which is usually checked into version control.

> Either way the secret leaks. Use the credential system instead, as shown in

Reviewer 2: ```suggestion
Either way the secret leaks. Use the [credential system]({{< relref "docs/concepts/credential-system.md" >}}) instead, as shown in
```

[Authenticate against a protected server](#authenticate-against-a-protected-server).
{{< /callout >}}

{{< /step >}}

{{< step >}}

### Build the component version

```bash

> ocm add cv --repository ./transport-archive --constructor component-constructor.yaml

Reviewer 2: I guess with our defaults, we can just do an `ocm add cv` and skip the params

Reviewer 3: True 👍

    - access:
        localReference: sha256:6e3205bbad194f902ee8bdc5a712c47f7a5443fabfd066ef1e95088c837fe0ae
        mediaType: application/x-tar
        type: LocalBlob/v1
      digest:
        hashAlgorithm: SHA-256
        normalisationAlgorithm: genericBlobDigest/v1
        value: 6e3205bbad194f902ee8bdc5a712c47f7a5443fabfd066ef1e95088c837fe0ae
      name: ocm-cli
      relation: local
      type: blob
      version: 1.0.0
```

{{< /details >}}

The **access type** keeps the `Wget/v1` specification as written:

{{< details "Expected output" >}}

```yaml
    resources:
    - access:
        mediaType: application/x-tar
        type: Wget/v1
        url: https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar
      digest:
        hashAlgorithm: SHA-256
        normalisationAlgorithm: genericBlobDigest/v1
        value: 6e3205bbad194f902ee8bdc5a712c47f7a5443fabfd066ef1e95088c837fe0ae
      name: ocm-cli
      relation: external
      type: blob
      version: 1.0.0
```

{{< /details >}}

In both cases, the resource has the same digest, computed over the bytes that were fetched.

{{< /step >}}

{{< step >}}

### Download the resource back

> This proves the transport (and, on a protected server, the credentials) work end to end:

Reviewer 2: ```suggestion
This proves the transport (and, on a protected server, the credentials) work end-to-end:
```

```bash
ocm download resource ./transport-archive//github.com/acme.org/myapp:1.0.0 \
  --identity name=ocm-cli \
  --output ./ocm-cli
```

You should see: `level=INFO msg="resource downloaded successfully" output=./ocm-cli`.

{{< callout context="note" >}}
When the media type is an archive the CLI can unpack, `--output` is a **directory** with the extracted contents.
Other media types are written to a file at that path. `cli.tar` is a `.tar`, so `./ocm-cli` is a directory here.
{{< /callout >}}

{{< /step >}}

{{< /steps >}}

## Authenticate against a protected server

> The release asset above is public. When the URL is behind authentication (an artifact repository, or a private

Reviewer 2: ```suggestion
The release asset above is public. When accessing the URL requires authentication (an artifact repository, or a private
```

> GitHub asset), configure the credentials instead of putting them in the specification. OCM matches them to the request

Reviewer 2: ```suggestion
GitHub asset), [configure credentials]({{< relref "docs/how-to/configure-multiple-credentials.md" >}}) of type `WgetCredentials` instead of putting them in the specification. OCM matches them to the request
```

by a consumer identity of type `Wget` derived from the URL.

Write the credentials to `.ocmconfig`:

```bash
cat > .ocmconfig << 'EOF'
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: Wget
          hostname: downloads.example.com

>           scheme: https

Reviewer 4: Isn't the consumer identity type versioned? I guess not since we do not have typed consumer identities yet, right?

Reviewer 3: Do you want me to do anything about that here?

        credentials:
          - type: WgetCredentials/v1
            username: my-user
            password: my-password
EOF
```

> Then pass `--config` when you build (or drop the file in one of the

Reviewer 2: ```suggestion
Then pass `--config` when you build the component (or drop the file in one of the
```

[well-known locations]({{< relref "docs/reference/ocm-cli/ocm.md" >}}) the CLI reads automatically):

```bash
ocm --config .ocmconfig add cv \
  --repository ./transport-archive \

>   --constructor component-constructor.yaml

Reviewer 4: same as mentioned above, `--repository` and `--constructor` could be omitted

```

HTTP Basic Auth, bearer tokens, and mutual TLS are all supported. For how the identity is matched and how the auth

> methods interact, see [Authentication]({{< relref "docs/tutorials/wget-http-resources.md#authentication" >}}) in the

Reviewer 2: ```suggestion
methods interact, see the [Authentication]({{< relref "docs/tutorials/wget-http-resources.md#authentication" >}}) section in the
```

This guide fetches a public OCM release asset, so it runs end-to-end with no registry and no credentials. To point it
at a server that needs authentication, see [Authenticate against a protected server](#authenticate-against-a-protected-server).

## Steps

{{< steps >}}

{{< step >}}

### Create the component constructor

Both variants below describe the same resource with the same fields. Write one of them to `component-constructor.yaml`:

{{< tabs "wget-spec" >}}
{{< tab "Input type" >}}

```bash
cat > component-constructor.yaml << 'EOF'
components:
  - name: github.com/acme.org/myapp
    version: 1.0.0
    provider:
      name: acme.org
    resources:
      - name: ocm-cli
        type: blob
        version: 1.0.0
        input:
          type: Wget/v1
          url: https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar
          mediaType: application/x-tar
EOF
```

{{< /tab >}}
{{< tab "Access type" >}}

```bash
cat > component-constructor.yaml << 'EOF'
components:
  - name: github.com/acme.org/myapp
    version: 1.0.0
    provider:
      name: acme.org
    resources:
      - name: ocm-cli
        type: blob

>         version: 1.0.0

Reviewer 5: 0.12, let's make it a bit realistic:)

Reviewer 3: Technically, the resource version can be separate from the CLI. :) But yes. I will update these things anyways because I need to refactor the entire tutorial. :)

[Authenticate against a protected server](#authenticate-against-a-protected-server).
{{< /callout >}}

{{< /step >}}

{{< step >}}

### Build the component version

```bash
ocm add cv
```

The asset is ~25 MB, so this takes a moment. You should see the component listed:

```text
 COMPONENT                 │ VERSION │ PROVIDER
───────────────────────────┼─────────┼──────────
 github.com/acme.org/myapp │ 1.0.0   │ acme.org
```

{{< /step >}}

{{< step >}}

### Check what ended up in the descriptor

```bash
ocm get cv ./transport-archive//github.com/acme.org/myapp:1.0.0 -o yaml
```

The **input type** stores the bytes locally, so the access becomes `LocalBlob/v1`:

{{< details "Expected output" >}}

```yaml
    resources:
    - access:
        localReference: sha256:6e3205bbad194f902ee8bdc5a712c47f7a5443fabfd066ef1e95088c837fe0ae
        mediaType: application/x-tar
        type: LocalBlob/v1
      digest:
        hashAlgorithm: SHA-256
        normalisationAlgorithm: genericBlobDigest/v1
        value: 6e3205bbad194f902ee8bdc5a712c47f7a5443fabfd066ef1e95088c837fe0ae
      name: ocm-cli
      relation: local
      type: blob

>       version: 1.0.0

Reviewer 5: 0.12 ?

|-------------------------------------------------------------|-----------------------------------------------------------------|
| The content must be reproducible and available offline      | The URL is the authoritative location and should stay so        |
| The upstream URL may disappear or change content            | The artifact is large and should not be duplicated              |
| You are building an air-gapped or self-contained repository | Consumers are expected to fetch directly from the origin server |

The distinction only holds for as long as the component version stays where it was built. Wget resources are always
transferred by value, so transferring a component version turns its `Wget/v1` access specifications into
[`LocalBlob/v1`]({{< relref "docs/reference/input-and-access-types.md#localblobv1" >}}) entries in the target
repository.

{{< /step >}}

{{< step >}}

### Add the resource to your constructor

{{< tabs "wget-spec" >}}
{{< tab "Input type" >}}

```yaml
resources:
  - name: release-archive
    type: blob
    version: 1.0.0
    input:
      type: Wget/v1

>       url: https://downloads.example.com/myapp/1.0.0/myapp-linux-amd64.tar.gz

Reviewer: I believe a real-life example would make sense. For example an URL to our releases `https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar`

| Custom CA       | Root CAs from credentials always applied        | `certificateAuthority` applied only alongside `certificate`     |

The precedence flip only becomes visible when a single entry carries `identityToken` **and** both `username` and
`password`: OCM v1 sends Basic Auth, OCM v2 sends the bearer token and logs a warning. With only a token, or only a
username/password pair, both versions behave identically. Split such entries if the server distinguishes the two
mechanisms.

## Troubleshooting

### Symptom: `401 Unauthorized` from the origin server

**Cause:** No consumer entry matched, so the request went out unauthenticated. The identity type is matched by exact
string and is unversioned, so neither `Wget/v1` nor the lowercase `wget` used by OCM v1 will match.
Another silent cause: an entry that sets only certificateAuthority without certificate. In OCM v2, certificateAuthority is only evaluated alongside a client certificate. It is not applied for server-side certificate verification on its own.

**Fix:** Set `type: Wget` in the consumer identity, and drop `pathprefix`. `path` is matched with `path.Match`, whose

> `*` does not cross `/`, so it cannot express a prefix over nested paths. Omitting it matches every path on the host:

Reviewer: I don't understand where `pathprefix` is coming from. If the symptom is `401 Unauthorized` it might be incorrect credentials or no identity was matching. In the latter case, I think I would rather point to the credential-reference section for `wget` and telling the user to make sure it matches instead of listing arbitrary failure-causes (at least it seems arbitrary to me :D)

Author: `pathprefix` is ocmv1 - since we mention the `ocmv1` diff and migration, I think this is worth listing

Author: split it into a general 401 and a ocmv1 migration solution

      body: eyJmb3JtYXQiOiJqc29uIn0=
```

{{< callout context="caution" >}}
**Never put credentials in `url`, `header`, or `body`.** That includes userinfo (`https://user:token@host/...`) and
presigned query parameters. Anything that is part of the specification is stored, not just headers.

An **access** specification is stored verbatim in the component descriptor: everything written there is persisted with
the component version, travels with it through every transfer, is covered by its signature, and is readable by anyone
who can read the component version. An **input** specification is resolved at construction time and never reaches the
descriptor, but it does live in your `component-constructor.yaml`, which is normally checked into version control.
Configure authentication through the credential system instead, as described in the next step.
{{< /callout >}}

{{< /step >}}

{{< step >}}

### Set the media type explicitly

The media type of the resulting blob is resolved in this order:

1. The `mediaType` field of the specification, if set.
2. The `Content-Type` header of the HTTP response.
3. `application/octet-stream`.

{{< callout context="note" >}}

> Unlike OCM v1, the media type is **not** derived from the URL's file extension. Set `mediaType` whenever the server does

Reviewer 2: I would remove the "Unlike in OCM v1..." sentence as this is covered in the migration table in line 382. The how-to targets V2 users, so the callout does not need to mention v1.

|-----------------|-------------------------------------------------|-----------------------------------------------------------------|
| Auth precedence | Basic Auth wins; the bearer token is a fallback | Bearer token wins; Basic Auth is used only when no token is set |
| Basic Auth      | Requires both `username` and `password`         | Sent as soon as `username` is set                               |
| Custom CA       | Root CAs from credentials always applied        | `certificateAuthority` applied only alongside `certificate`     |

The precedence flip only becomes visible when a single entry carries `identityToken` **and** both `username` and
`password`: OCM v1 sends Basic Auth, OCM v2 sends the bearer token and logs a warning. With only a token, or only a
username/password pair, both versions behave identically. Split such entries if the server distinguishes the two
mechanisms.

## Troubleshooting

### Symptom: `401 Unauthorized` from the origin server

**Cause:** No consumer entry matched, so the request went out unauthenticated. The identity type is matched by exact
string and is unversioned, so neither `Wget/v1` nor the lowercase `wget` used by OCM v1 will match.

> 

Reviewer 2: ```suggestion
Another silent cause: an entry that sets only certificateAuthority without certificate. In OCM v2, certificateAuthority is only evaluated alongside a client certificate. It is not applied for server-side certificate verification on its own.

```

    version: 1.0.0
    provider:
      name: acme.org
    resources:
      - name: ocm-cli
        type: blob
        version: 1.0.0
        input:
          type: Wget/v1
          url: https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar
          mediaType: application/x-tar
EOF
```

{{< /tab >}}
{{< tab "Access type" >}}

```bash
cat > component-constructor.yaml << 'EOF'
components:
  - name: github.com/acme.org/myapp
    version: 1.0.0
    provider:
      name: acme.org
    resources:
      - name: ocm-cli
        type: blob
        version: 1.0.0
        relation: external
        access:
          type: Wget/v1
          url: https://github.com/open-component-model/open-component-model/releases/download/v0.12.0/cli.tar
          mediaType: application/x-tar
EOF
```

{{< /tab >}}
{{< /tabs >}}

`mediaType` is set explicitly because GitHub serves every release asset as `application/octet-stream`. See
[Setting the media type](#setting-the-media-type) for the full resolution order.

{{< callout context="caution" >}}
**Never put credentials in `url`, `header`, or `body`.** That includes userinfo (`https://user:token@host/...`) and

> presigned query parameters. Anything that is part of the specification is stored, not just headers.

Reviewer: What is meant by "not just headers"?

An **access** specification is stored verbatim in the component descriptor: everything written there is persisted with
the component version, travels with it through every transfer, is covered by its signature, and is readable by anyone

> who can read the component version. An **input** specification is resolved at construction time and never reaches the

Reviewer 2: ```suggestion
who can read the component version.  

An **input** specification is resolved at construction time and never reaches the
```

Reviewer 2: I would add a new line in between input an access type. The callout is anyway quite long so should be fine.

> descriptor, but it does live in your `component-constructor.yaml`, which is normally checked into version control.

Reviewer 2: ```suggestion
descriptor, but it does live in your `component-constructor.yaml`, which is normally checked into version control.  

```

Reviewer 2: same here. The generic part with the authentication config should be split from the access type section.

      digest:
        hashAlgorithm: SHA-256
        normalisationAlgorithm: genericBlobDigest/v1
        value: 6e3205bbad194f902ee8bdc5a712c47f7a5443fabfd066ef1e95088c837fe0ae
      name: ocm-cli
      relation: external
      type: blob
      version: 1.0.0
```

{{< /details >}}

Either way the resource carries the same digest, computed over the bytes that were fetched.

{{< /step >}}

{{< step >}}

### Download the resource back

Downloading proves the transport and, on a protected server, the credentials work end to end:

```bash
ocm download resource ./transport-archive//github.com/acme.org/myapp:1.0.0 \
  --identity name=ocm-cli \
  --output ./ocm-cli
```

You should see: `level=INFO msg="resource downloaded successfully" output=./ocm-cli`.

{{< callout context="note" >}}
When the media type is an archive format the CLI can unpack, `--output` is a **directory** holding the extracted
contents. Other media types are written to a file at that path. `cli.tar` is a `.tar`, so `./ocm-cli` is a directory
here; its contents happen to be an OCI image layout, because that is what this particular release asset packages.
{{< /callout >}}

{{< /step >}}

{{< /steps >}}

> ## Setting the media type

Reviewer: This could also rather be a callout than its own section

The media type of the resulting blob is resolved in this order:

1. The `mediaType` field of the specification, if set.
2. The `Content-Type` header of the HTTP response.
3. `application/octet-stream`.

Set `mediaType` whenever the server does not send a useful `Content-Type`. Otherwise the resource ends up as
`application/octet-stream`. GitHub release assets are a common case: they are served as `application/octet-stream`
regardless of what they contain, which is why the walkthrough sets `mediaType: application/x-tar` explicitly.

> ## Authenticating against a protected server

Reviewer: IMO we should not have a whole section about this because (a) it looks like its own little how-to and (b) it looks like the usual credential handling of ocm. I would just add a call-out and refer the respective `wget` credential reference section.

        type: blob
        version: 1.0.0
        relation: external
        access:
          type: Wget/v1
          url: https://downloads.example.com/myapp/1.0.0/myapp-linux-amd64.tar.gz
          mediaType: application/x-gzip
EOF
```

Then pass the configuration explicitly, or leave it in one of the
[well-known locations]({{< relref "docs/reference/ocm-cli/ocm.md" >}}) where the CLI picks it up automatically:

```bash
ocm --config .ocmconfig add cv \
  --repository ./transport-archive \
  --constructor component-constructor.yaml
```

HTTP Basic Auth, bearer tokens, and mutual TLS are supported. Basic Auth and a bearer token both set the
`Authorization` header and are mutually exclusive. When both are configured, the bearer token wins and a warning is
logged. A client certificate combines with either, but only takes effect during a TLS handshake.

See [Credential Consumer Identities: Wget]({{< relref "docs/reference/credential-consumer-identities.md#wget" >}}) for
the identity attributes and matching rules, and
[Credential Types: WgetCredentials/v1]({{< relref "docs/reference/credential-types.md#wgetcredentialsv1" >}}) for the
full field reference.

> ## Pinning an expected digest

Reviewer: I believe it would make sense to put this above in the How-To. This is a best-practice that users should be aware of (putting it down here, makes it likely to be skipped)
